Grbl_Esp32/src/GCode.cpp

```diff
@@ -1301,31 +1420,40 @@ uint8_t gc_execute_line(char* line, uint8_t client) {
     // M0,M1,M2,M30: Perform non-running program flow actions. During a program pause, the buffer may
     // refill and can only be resumed by the cycle start run-time command.
     gc_state.modal.program_flow = gc_block.modal.program_flow;
-    if (gc_state.modal.program_flow) {
-        protocol_buffer_synchronize();  // Sync and finish all remaining buffered motions before moving on.
-        if (gc_state.modal.program_flow == PROGRAM_FLOW_PAUSED) {
+    switch (gc_state.modal.program_flow) {
+        case ProgramFlow::Running: break;
+        case ProgramFlow::OptionalStop:
```

Comment:
That is intentional.  In Running state (==0), the existing code did not synchronize - which is correct since you want the planner/runner to keep working ahead of the parser.  The OptionalStop case is where the behavior differs - but OptionalStop is rejected at an earlier place in the code so this will not be reached.  If we do support OptionalStop later, as mentioned in the TODO in the next line, then the code that chooses whether or not to stop will break if no stop, and thus should not synchronize, or else fall through to the Paused case, which will then synchronize.  So I changed a "can't happen, but would not do the right thing if it did happen", to "would do the right thing with the addition of code to evaluate whether or not to stop".  Now that we have Settings and could easily add a "whether to stop" one, we could consider supporting M1.